Turn off spinner if spread doesn't begin loading

ettore opened the pull request:
If the spread does begin to load, it will send a `spreadLoadStarted` event that will keep the spinner spinning.

This may happen if for example there's a broken/corrupted DRM key: this will cause content decryption to fail and therefore the spread won't be able to load.

This solves Issue #396 .

ettore commented:
I left quite a few debug statements in, but I thought they would be helpful to document what's going on. Let me know if I should remove some.

ettore commented on Sources/Navigator/EPUB/EPUBSpreadView.swift:
```diff
@@ -522,6 +536,33 @@ private extension EPUBSpreadView {
+        // If the spread begins to load it will send a `spreadLoadStart` JS
+        // event which will cancel the work item being scheduled here.
+        log(.debug, "\(type(of: self))::\(#function) scheduling activityIndicatorStopWorkItem")
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1,
```
this 1 sec is obviously arbitrary, but in a way that is not hurtful. It just means that we are leaving up to 1 second for the spread to begin loading before stopping the spinner. So, I don't believe it's too long. The spinner can and will be stopped before 1 second in case of no errors.

If it's too short, it just means that the spread takes a really long time to begin loading for some reason. In that case the user might see an interval with no spinner.

ettore commented on Sources/Navigator/EPUB/Scripts/src/index-fixed.js:
```diff
@@ -9,3 +9,7 @@
 window.readium.isFixedLayout = true;
+
+window.addEventListener("load", function () {
+  webkit.messageHandlers.spreadLoadStarted.postMessage({});
```
this works even if I put the `postMessage` call outside of the event listener. I don't know what's best. I simply emulated what's being done on the reflowable case.

ettore replied:
on the reflowable case though we already have a `load` event we are listening to. Wouldn't that mark the beginning of the spread loading more accurately than just putting it outside the `window.addEventListener` call?  

Or is it that the simple fact that `index-reflowable.js` is being executed the mark that the spread is starting to load?

mickael-menu replied:
> Or is it that the simple fact that `index-reflowable.js` -- and `index-fixed.js` -- are being executed the mark that the spread is starting to load?

Yes this one. If we use the `window`'s `load` event, it will wait for all the HTML being loaded, as well as the external resources (stylesheets, images, etc.). Here we want to get the `spreadLoadStarted` as soon as possible to make sure we don't hide the spinner too early.

ettore replied:
fixed in both places.

ettore commented on Sources/Navigator/EPUB/EPUBSpreadView.swift:
```diff
@@ -228,12 +229,18 @@ class EPUBSpreadView: UIView, Loggable, PageView {
     }
 
+    private func spreadLoadDidStart(_ body: Any) {
+        log(.debug, "\(type(of: self))::\(#function)")
```
definitely, I can do that. 👍 

I think the trace level logging could be useful especially when debugging code with delays and just asynchronous events in general. At least it has been to me, because putting breakpoints in those situations alters the lifecycle of the web view and its components.

ettore commented on Sources/Shared/Logger/Loggable.swift:
```diff
@@ -94,3 +99,11 @@ public extension Loggable {
+
+public extension Loggable {
+    func trace(_ message: String = "", function: String = #function, file: String = #file, line: Int = #line) {
+        #if DEBUG
```
I added this #ifdef because it seems like not only we will never want `trace` level statements in a production build, but also we won't even want the performance hit of executing the log function code: for example, the `Logger` class has an `if` that will always be executed no matter what logging level the user chooses. I think this forceful behavior is particular of "trace" level semantics, but let me know if you disagree.

mickael-menu replied:
Good call, thank you 👌

ettore commented:
not sure why the Carthage integration checks always fail... 🤔

ettore commented:
I just added a commit to handle some `yarn` generated files. After some research, I have added some of them to source control and git-ignored others. Let me know if you have any preferences here.

ettore commented on .editorconfig:
I have seen arguments for ignoring this and others to add it. But it appears that this file aims to suggest proper recommendations for linting, so it seems appropriate to check it in. Let me know if you disagree or have any preferences.

mickael-menu replied:
I'm fine with it but would add it under `Sources/Navigator/EPUB/Scripts`?

ettore commented on yarn.lock:
according to official docs this should always be committed: https://classic.yarnpkg.com/blog/2016/11/24/lockfiles-for-all/

ettore commented on package.json:
package.json is usually committed, but I'm unsure if this is just because of my developer setup. If you don't have it, I should probably gitignore it.

mickael-menu commented on Sources/Shared/Logger/Loggable.swift:
```diff
@@ -19,6 +20,8 @@ public enum SeverityLevel: String {
     var symbol: String {
         switch self {
+        case .trace:
+            return "🐾"
```
Haha perfect choice!

mickael-menu commented on Sources/Shared/Logger/Loggable.swift:
```diff
@@ -19,6 +20,8 @@ public enum SeverityLevel: String {
     var symbol: String {
         switch self {
+        case .trace:
+            return "🐾"
```
Haha perfect choice!

ettore replied:
one of my proudest achievements 😄

mickael-menu reviewed (commented):
> not sure why the Carthage integration checks always fail... 🤔

Don't worry about this one. It works only when the PR is made on a branch on the repo. Unfortunately I didn't find a way to work around this limitation with Carthage yet.

> I just added a commit to handle some yarn generated files. After some research, I have added some of them to source control and git-ignored others. Let me know if you have any preferences here.

I'm a bit confused about why the files were added to the root of the repo. We already have these files (although with #392 we're using `pnpm-lock.yaml` instead of `yarn.lock`) but in the folder containing the scripts (`Sources/Navigator/EPUB/Scripts`).

Did you use `make scripts` from the root of the repo, to regenerate the JS files?

ettore commented:
> Did you use `make scripts` from the root of the repo, to regenerate the JS files?

I did... it seems like I was supposed to run it from `Sources/Navigator/EPUB/Scripts`? if so I can remove them and run the scripts from there.

ettore commented:
> > Did you use `make scripts` from the root of the repo, to regenerate the JS files?
> 
> I did... it seems like I was supposed to run it from `Sources/Navigator/EPUB/Scripts`? if so I can remove them and run the scripts from there.

I just tried to run `make scripts` from inside `Sources/Navigator/EPUB/Scripts` but there's no `Makefile` there.
```
ettorepasquini@cube25 (fix/neverending-loading-state +$<) ~/git/swift-toolkit/Sources/Navigator/EPUB/Scripts: make scripts
make: *** No rule to make target `scripts'.  Stop.
```
The `Makefile` is at the root of the repo; not sure if that should be moved to the `Sources/Navigator/EPUB/Scripts` dir?

mickael-menu commented:
> I did... it seems like I was supposed to run it from Sources/Navigator/EPUB/Scripts? if so I can remove them and run the scripts from there.

No it's made to be run from the root. [It automatically moves to `Sources/Navigator/EPUB/Scripts`](https://github.com/readium/swift-toolkit/blob/0573d3296ffbca7aa78cbf98d97c4461d7fbc30f/Makefile#L20).

I'm confused why it generated these files. I tried from a fresh clone and didn't have the issue:

```shell
git clone git@github.com:readium/swift-toolkit.git
cd swift-toolkit
make scripts
```

I think you can just remove these files for the PR, the rest looks fine.

mickael-menu reviewed (approved):
Thank you for making the changes @ettore 🙏
